### PR TITLE
Detect direct/internal usage of all missing library symbols

### PR DESCRIPTION
**Pull request by sbc100**
Previously this mechanism only worked for missing function symbols and
would only fire once the function was called.  The new mechanism will
catch any access to the missing symbol, even if its never called.

The downside of this change is that in debug builds (when ASSERTIONS are
enabled) we now call `missingLibrarySymbol` on a lot more symbols.
However, I also optimized as part of this change such that the code for
calling `missingLibrarySymbol` is not a lot smaller.

Old:

```
missingLibrarySymbol('foo');
missingLibrarySymbol('bar');
...
```

New:

```
var missingLibrarySymbols = ['foo', 'bar' ...];
missingLibrarySymbols.forEach(missingLibrarySymbol);
```

### CODE REVIEW

File: src/modules.js [hunk 1]
```diff
@@ -336,14 +343,6 @@ function exportRuntime() {
       }
       return `Module["${name}"] = ${exported};`;
     }
-    // do not export it. but if ASSERTIONS, emit a
-    // stub with an error, so the user gets a message
-    // if it is used, that they should export it
-    if (ASSERTIONS) {
-      // check if it already exists, to support EXPORT_ALL and other cases
-      const fssymbol = isExportedByForceFilesystem(name);
```
Comment (kripken) [inline src/modules.js:344]: This function has no more uses.

File: src/runtime_debug.js [hunk 2]
```diff
@@ -22,24 +22,26 @@ function ignoredModuleProp(prop) {
   }
 }
 
-function unexportedMessage(sym, isFSSybol) {
-  var msg = "'" + sym + "' was not exported. add it to EXPORTED_RUNTIME_METHODS (see the FAQ)";
-  if (isFSSybol) {
-    msg += '. Alternatively, forcing filesystem support (-sFORCE_FILESYSTEM) can export this for you';
```
Comment (kripken) [inline src/runtime_debug.js:28]: Looks like we lose the specific filesystem integration here? Maybe not a big loss.

File: tests/test_core.py [hunk 3]
```diff
@@ -7034,23 +7034,33 @@ def test(output_prefix='', args=None, assert_returncode=0):
...
-    self.set_setting('ASSERTIONS')
-    test('_assert', assert_returncode=NON_ZERO)
+    # Without assertion indirect usages (via Module) result in `undefined` and direct usage
+    # generates a builtin (not very helpful) JS error.
     self.set_setting('ASSERTIONS', 0)
-    # see that when we export them, things work on the module
+    test('|undefined|')
+    test('ALLOC_STACK is not defined', args=['-DDIRECT'], assert_returncode=NON_ZERO)
+
+    # When assertions are enabled direct and indiect usage both abort with a useful error message.
```
Comment (kripken) [inline tests/test_core.py:7047]:
```suggestion
    # When assertions are enabled direct and indirect usage both abort with a useful error message.
```

File: src/runtime_debug.js (hunk 2 above)
Comment (sbc100) [reply src/runtime_debug.js:28]: Reinstated;

File: src/library.js [hunk 4]
```diff
@@ -3684,6 +3684,7 @@ DEFAULT_LIBRARY_FUNCS_TO_INCLUDE.push(
   '$addFunction',
   '$removeFunction',
   '$allocate',
+  '$ALLOC_STACK',
```
Comment (kripken) [inline src/library.js:3687]: Why just `ALLOC_STACK` and not the other `ALLOC_`?

Comment (sbc100) [reply src/library.js:3688]: Good point, I added the other. I just included that one because it was used in test.

**Pull request merged**